common/clock/ratelimiter.go

```diff
@@ -0,0 +1,418 @@
+		// usable token, return a cancel-able handler in case it's not used
+		return &allowedReservation{
+			res:        res,
+			reservedAt: now,
+			limiter:    r,
+		}
+	}
+	// unusable, immediately cancel it to get rid of the future-reservation
+	// since we don't allow waiting for it.
+	//
+	// this restores MANY more tokens when heavily contended, as time skews
+	// less before canceling, so it gets MUCH closer to the intended limit.
+	res.CancelAt(now)
+	return deniedReservation{}
+}
+
+func (r *ratelimiter) Burst() int {
+	// rate.Limiter.Burst does not currently advance time,
+	// so this does not need to hold the ratelimiter lock.
+	// if this changes, use lockNow.
+	return r.limiter.Burst()
+}
+
+func (r *ratelimiter) Limit() rate.Limit {
+	// rate.Limiter.Limit does not currently advance time,
+	// so this does not need to hold the ratelimiter lock.
+	// if this changes, use lockNow.
+	return r.limiter.Limit()
+}
+
+func (r *ratelimiter) SetBurst(newBurst int) {
+	now, unlock := r.lockNow()
+	defer unlock()
+	r.limiter.SetBurstAt(now, newBurst)
+}
+
+func (r *ratelimiter) SetLimit(newLimit rate.Limit) {
+	now, unlock := r.lockNow()
+	defer unlock()
+	r.limiter.SetLimitAt(now, newLimit)
+}
+
+func (r *ratelimiter) Tokens() float64 {
+	now, unlock := r.lockNow()
+	defer unlock()
+	return r.limiter.TokensAt(now)
+}
+
+func (r *ratelimiter) Wait(ctx context.Context) (err error) {
+	now, unlock := r.lockNow()
+	var once sync.Once
+	defer once.Do(unlock) // unlock if panic or returned early with no err
+
+	res := r.limiter.ReserveN(now, 1)
+	defer func() {
+		if err != nil {
+			// err return means "not allowed", so cancel the reservation.
+			//
+			// note that this makes a separate call to get the current time:
+			// this is intentional, as time may have passed while waiting.
+			//
+			// if the cancellation happened before the time-to-act (the delay),
+			// the reservation will still be successfully rolled back.
+
+			// ensure we are unlocked, which will not have happened if an err
+			// was returned immediately.
+			once.Do(unlock)
+
+			// (re)-acquire the latest now value.
+			//
+			// it should not be advanced to the "real" now, to improve the chance
+			// that the token will be restored, but if there was any waiting
+			// it's pretty likely that other calls occurred while unlocked and
+			// the most-recently-observed time must be maintained.
+			now, unlock := r.lockReservedNow(now)
+			defer unlock()
+			res.CancelAt(now)
+		}
+	}()
+
+	if !res.OK() {
+		// !OK can only mean "impossible to wait" with `ReserveN`.
+		// Since there is no deadline passed to the limiter (the reservation
+		// contains the delay), and we only ever request 1 token, this can only
+		// mean a burst of 0 (and non-infinite rate.Limit).
+		return errWaitZeroBurst
+	}
+	delay := res.DelayFrom(now)
+	if delay == 0 {
+		return nil // available token, allow it
+	}
+	if deadline, ok := ctx.Deadline(); ok && now.Add(delay).After(deadline) {
```

Comment:
We may want a few milliseconds of buffer (now+delay < deadline-buffer). Ideally specified by the endpoint being protected but due to lack of annotations that will be a mess. Probably fine to leave like this for now

Reply:
from IRL: this kind of "reserve a buffer for the caller to use the thing protected by the limiter" is better handled by layers above this, by making a context with a deadline before that buffer.

which does very much seem worth making easier to do, it's quite verbose with plain contexts and I suspect that's part of why we don't do it much.  and it may be worth building a `Wait(ctx, reserveTimeBeforeCtxTimesOut)` API to make it super easy and clear.  we just don't really have anything that works this way right now, and it wouldn't be implemented at this layer anyway.